subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationRole.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.configurations;
+
+import org.gradle.internal.deprecation.DeprecationLogger;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Defines how a {@link org.gradle.api.artifacts.Configuration} is intended to be used.
+ *
+ * Standard roles are defined in {@link ConfigurationRoles}.
+ *
+ * @since 8.0
+ */
+public interface ConfigurationRole {
+    /**
+     * Returns a human-readable name for this role.
+     */
+    String getName();
+
+    boolean isConsumable();
+    boolean isResolvable();
+    boolean isDeclarableAgainst();
+    boolean isConsumptionDeprecated();
+    boolean isResolutionDeprecated();
+    boolean isDeclarationAgainstDeprecated();
+
+    /**
+     * Obtains a human-readable summary of the usage allowed by the given role.
+     */
+    default String describe() {
```

Comment:
How come we don't just implement `toString` here? Why do we need to invent a new method for this?

Reply:
This isn't meant to be `toString` which would include every property and perhaps additional fields on an implementation class.  It's just meant to describe the allowed usage.  I'll rename the method to make this more clear.

Reply:
No need to rename, I think the name is clear enough. What's the downside of having all that info in the toString?